### PR TITLE
Shunt to work with HEAD container for migrations

### PR DESCRIPTION
**Pull request by mdegerne**
Specifically, when a container has not been migrated yet, the headers from
the source should be displayed.

### CODE REVIEW

File: s3_sync/shunt.py [hunk 1]
```diff
@@ -391,6 +395,40 @@ def handle_listing(self, req, start_response, sync_profile, cont,
...
+        utils.close_if_possible(app_iter)
+
+        provider = create_provider(sync_profile, max_conns=1,
+                                   per_account=per_account)
+
+        resp = provider.head_bucket(sync_profile['aws_bucket'])
+        if resp.status != 200:
+            # return original failure
+            start_response(status, headers)
+            return app_iter
```
Comment (dbishop) [inline s3_sync/shunt.py]: You closed app_iter (if possible), above.  This is probably going to fail here?

File: s3_sync/shunt.py [hunk 2]
```diff
@@ -391,6 +395,40 @@ def handle_listing(self, req, start_response, sync_profile, cont,
...
+
+        self.logger.debug("MSD: remote-headers: %s" % resp.headers)
```
Comment (dbishop) [inline s3_sync/shunt.py]: Was this intended to be left in?

**Comment by mdegerne** [comment]
Both valid complaints. Working....

File: s3_sync/shunt.py [hunk 3]
```diff
@@ -391,6 +395,37 @@ def handle_listing(self, req, start_response, sync_profile, cont,
         start_response(status, dict_headers.items())
         return response
 
+    def handle_head(self, req, start_response, sync_profile, cont,
```
Comment (timuralp) [inline s3_sync/shunt.py]: I'd rename this to `container_head` to differentiate from object HEAD.

File: s3_sync/shunt.py [hunk 4]
```diff
@@ -260,6 +260,10 @@ def __call__(self, env, start_response):
         if not obj and req.method == 'GET':
             return self.handle_listing(req, start_response, sync_profile, cont,
                                        per_account)
+        elif not obj and req.method == 'HEAD' and \
```
Comment (timuralp) [inline s3_sync/shunt.py]: Consider merging this with the prior if clause, which would be:
```
if not obj:
    if req.method == 'GET':
        ...
    elif req.method == 'HEAD':
        ...
```

This makes it obvious we're handling container requests.

File: test/integration/test_migrator.py [hunk 5]
```diff
@@ -385,12 +385,21 @@ def _check_container_created(conn):
                     return False
                 raise
 
+        # verify get_head works through shunt
+        hdrs = conn_local.head_container(migration['container'])
+        self.assertIn('x-container-meta-test', hdrs)
+        self.assertEqual('test metadata', hdrs['x-container-meta-test'])
```
Comment (timuralp) [inline test/integration/test_migrator.py]: Consider testing with unicode container metadata.

File: s3_sync/shunt.py [hunk 6]
```diff
@@ -391,6 +395,37 @@ def handle_listing(self, req, start_response, sync_profile, cont,
...
+
+        headers = [(k.encode('utf-8'), unicode(v).encode('utf-8'))
+                   for k, v in resp.headers.iteritems()]
+        self.logger.debug('Remote resp: %s' % status)
+
+        headers = filter_hop_by_hop_headers(headers)
+        headers.extend(trans_id_headers)
+
+        start_response('200 OK', headers)
+        return app_iter
```
Comment (timuralp) [inline s3_sync/shunt.py]: Same thing with `app_iter` that Darrell pointed out above.

File: s3_sync/shunt.py [hunk 7]
```diff
@@ -391,6 +395,37 @@ def handle_listing(self, req, start_response, sync_profile, cont,
...
+
+        resp = provider.head_bucket(sync_profile['aws_bucket'])
+        if resp.status != 200:
+            # return original failure
+            start_response(status, headers)
+            return app_iter
+
+        headers = [(k.encode('utf-8'), unicode(v).encode('utf-8'))
+                   for k, v in resp.headers.iteritems()]
+        self.logger.debug('Remote resp: %s' % status)
```
Comment (timuralp) [inline s3_sync/shunt.py]: Remote response is `resp.status`? I suspect copying/pasting code may be the issue.

File: s3_sync/shunt.py [hunk 8]
```diff
@@ -391,6 +395,37 @@ def handle_listing(self, req, start_response, sync_profile, cont,
...
+
+        headers = filter_hop_by_hop_headers(headers)
```
Comment (timuralp) [inline s3_sync/shunt.py]: We probably want to do it for all of our requests. I wonder if we can surface this to `__call__`.

**Comment by mdegerne** [comment]
Fixed issues listed and rebased/merged commits. Note that the unicode test is the later test.

File: s3_sync/shunt.py [hunk 9]
```diff
@@ -395,6 +399,37 @@ def handle_listing(self, req, start_response, sync_profile, cont,
...
+
+        headers = [(k.encode('utf-8'), unicode(v).encode('utf-8'))
+                   for k, v in resp.headers.iteritems()]
+        self.logger.debug('Remote resp: %s' % resp.status)
+
+        headers = filter_hop_by_hop_headers(headers)
+        headers.extend(trans_id_headers)
+
+        start_response('200 OK', headers)
+        return app_iter
```
Comment (dbishop) [inline s3_sync/shunt.py]: Wait, is this returning `app_iter` which is from the local cluster, or the body of the response from the provider (remote cluster)?

File: s3_sync/shunt.py [hunk 10]
```diff
@@ -395,6 +399,37 @@ def handle_listing(self, req, start_response, sync_profile, cont,
...
+            return app_iter
+
+        headers = [(k.encode('utf-8'), unicode(v).encode('utf-8'))
+                   for k, v in resp.headers.iteritems()]
+        self.logger.debug('Remote resp: %s' % resp.status)
+
+        headers = filter_hop_by_hop_headers(headers)
+        headers.extend(trans_id_headers)
+
+        start_response('200 OK', headers)
```
Comment (timuralp) [inline s3_sync/shunt.py]: 200 OK seems incorrect here (although, part of this is the fault of the provider API and S3). Swift returns 204 No Content on a container GET. S3 returns 200 OK. We should return 204 No Content on success.

File: s3_sync/shunt.py (hunk 9 above)
Comment (timuralp) [reply s3_sync/shunt.py]: On success, this should be an empty body. Although, even on a 404, Swift seems to return an empty body, so we might be ok.

File: test/integration/test_migrator.py [hunk 11]
```diff
@@ -444,17 +458,27 @@ def _check_objects_copied():
...
         for cont in test_containers:
-            self.swift_nuser.put_container(cont)
+            conn_remote.put_container(cont)
             for name, body, headers in test_objects:
-                self.nuser_swift('put_object', cont, name,
-                                 StringIO.StringIO(body), headers=headers)
+                conn_remote.put_object(
+                    cont, name, StringIO.StringIO(body), headers=headers)
+
+        hn = u'x-container-meta-tes\u00e9t'
```
Comment (timuralp) [inline test/integration/test_migrator.py]: You don't have to address this, but I prefer somewhat clearer variable names.

File: s3_sync/shunt.py (hunk 9 above)
Comment (dbishop) [reply s3_sync/shunt.py]: Oh, sorry, I forgot this was a HEAD handling function.  IIRC, the spec says HEAD bodies always have zero bytes (specifically, `...the server MUST NOT return a message-body in the response.`).

Comment (dbishop) [reply s3_sync/shunt.py]: ...that said, it seems like returning an empty body here would be more obviously correct (there's no _guarantee_ either of the other two responses correctly implemented HEAD and didn't include a message-body in their response...)

File: s3_sync/shunt.py (hunk 1 above)
Comment (timuralp) [reply s3_sync/shunt.py]: Done

File: test/integration/test_migrator.py (hunk 5 above)
Comment (timuralp) [reply test/integration/test_migrator.py]: Done.

File: s3_sync/shunt.py (hunk 6 above)
Comment (timuralp) [reply s3_sync/shunt.py]: Done -- returns [] now.

File: s3_sync/shunt.py (hunk 9 above)
Comment (timuralp) [reply s3_sync/shunt.py]: Done.

File: s3_sync/shunt.py (hunk 10 above)
Comment (timuralp) [reply s3_sync/shunt.py]: Done.

File: test/integration/test_migrator.py (hunk 11 above)
Comment (timuralp) [reply test/integration/test_migrator.py]: I changed this, as I think it does improve readability.

**Comment by timuralp** [comment]
Ugh -- this is swiftclient screwing with us. It mutates the header dictionary that we pass it :(

Fix will be up shortly.

**Comment by tipabu** [comment]
FWIW -- https://review.openstack.org/#/c/498069/ -- Stop mutating header dicts

**Pull request merged**